### PR TITLE
fix some small rtttl issues

### PR DESCRIPTION
**Pull request by nielsnl68**
# What does this implement/fix?

I got some error while playing the **H** note.
And added some better handling of Starting and finishing the Speaker using the Finishing process introduced by @glmnet. This makes sure the last note(s) is/are played correctly.

~~This PR is a follow up of pr #6718 and should only be merged when the later is being merged as well~~.

**NOTE:** This PR needs #7255 

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3953

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

No specific changes.

```yaml
# Example config.yaml

external_components:
  - source:
      type: git
      url: https://github.com/gnumpi/esphome_audio
      ref: finish_on_timeout
    components: [ adf_pipeline, i2s_audio]
  - source: github://pr#6817
    components:
      - speaker
      - rtttl
    refresh: 0s    

# packages:
# status_led: !include status_led.yaml

i2s_audio:
  - id: i2s_dplx
    i2s_lrclk_pin: GPIO33
    i2s_bclk_pin: GPIO19
    access_mode: exclusive

adf_pipeline:
  - platform: i2s_audio
    type: audio_out
    id: adf_i2s_out
    i2s_audio_id: i2s_dplx
    i2s_dout_pin: GPIO22

    adf_alc: true
    alc_max: 0.6

    sample_rate: 16000
    bits_per_sample: 16bit
    channel: right
    fixed_settings: false

speaker:
  - platform: adf_pipeline
    id: adf_speaker
    keep_pipeline_alive: false
    pipeline:
      - self
      - adf_i2s_out

rtttl:
  speaker: adf_speaker
  id: my_rtttl
  gain: 0.65

api:
  on_client_connected:
      - delay: 5s
      - rtttl.play: "Intro:d=32,o=5,b=300:g#4,h4,d#4,2p." 

```
Please **NOTE** at the moment of posting i do not get any audio out of the speaker.

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by xNUTx** [comment]
Without this fix PR#6718 will cause my unit to throw a fit once I start the microphone (in this process the rtttl is used to have an audible start of the listening phase).

```
[23:42:29][D][switch:012]: 'Voice Assistant On/Off' Turning ON.
[23:42:29][D][switch:055]: 'Voice Assistant On/Off': Sending state ON
[23:42:29][D][rtttl:051]: Playing song scale_up
[23:42:29][D][rtttl:294]: play time old: 750 div: 25 new: 783150 31326
[23:42:29][D][esp-idf:000][speaker_task]: I (130515) I2S: DMA Malloc info, datalen=blocksize=512, dma_buf_count=4

[23:42:29][D][i2s_audio:072][speaker_task]: Installing driver : yes
[23:42:29][D][rtttl:294]: play time old: 750 div: 26 new: 768924 29574
[23:42:29][D][i2s_audio.speaker:163]: Started I2S Audio Speaker
[23:42:29][D][rtttl:294]: play time old: 750 div: 30 new: 790200 26340
[23:42:29][D][rtttl:294]: play time old: 750 div: 31 new: 770691 24861
[23:42:29][D][rtttl:294]: play time old: 750 div: 35 new: 774900 22140
[23:42:29][D][rtttl:294]: play time old: 750 div: 39 new: 768924 19716
[23:42:29][D][rtttl:294]: play time old: 750 div: 44 new: 773476 17579
[23:42:29][D][rtttl:294]: play time old: 750 div: 47 new: 779354 16582
[23:42:29][D][rtttl:176]: Playback finished
[23:42:29][D][ring_buffer:024]: Created ring buffer with size 2048
[23:42:29][D][micro_wake_word:399]: Resetting buffers and probabilities
[23:42:29][D][micro_wake_word:195]: State changed from IDLE to START_MICROPHONE
[23:42:29][D][micro_wake_word:107]: Starting Microphone
[23:42:29][D][micro_wake_word:195]: State changed from START_MICROPHONE to STARTING_MICROPHONE
[23:42:29][D][esp-idf:000]: I (130784) I2S: DMA Malloc info, datalen=blocksize=1024, dma_buf_count=4

[23:42:29][D][i2s_audio:072]: Installing driver : yes
[23:42:29][D][micro_wake_word:195]: State changed from STARTING_MICROPHONE to DETECTING_WAKE_WORD
[23:42:29][W][component:157]: Component i2s_audio.microphone set Warning flag: unspecified
[23:42:29][W][component:170]: Component i2s_audio.microphone cleared Warning flag
[23:42:29][W][component:157]: Component i2s_audio.microphone set Warning flag: unspecified
[23:42:29][W][component:170]: Component i2s_audio.microphone cleared Warning flag
[23:42:29][W][component:157]: Component i2s_audio.microphone set Warning flag: unspecified
[23:42:29][W][component:170]: Component i2s_audio.microphone cleared Warning flag
[23:42:29][W][component:157]: Component i2s_audio.microphone set Warning flag: unspecified
[23:42:29][W][component:170]: Component i2s_audio.microphone cleared Warning flag
[23:42:29][W][component:157]: Component i2s_audio.microphone set Warning flag: unspecified
[23:42:29][W][component:170]: Component i2s_audio.microphone cleared Warning flag
[23:42:29][W][component:157]: Component i2s_audio.microphone set Warning flag: unspecified
[23:42:29][W][component:170]: Component i2s_audio.microphone cleared Warning flag
[23:42:29][W][component:157]: Component i2s_audio.microphone set Warning flag: unspecified
[23:42:29][W][component:170]: Component i2s_audio.microphone cleared Warning flag
[23:42:29][W][component:157]: Component i2s_audio.microphone set Warning flag: unspecified
[23:42:29][W][component:170]: Component i2s_audio.microphone cleared Warning flag
```
This last notification will keep coming until I stop the mww/microphone listening.

Just to iterate: it _FIXED_ my issue.

**Review by jesserockz — CHANGES_REQUESTED** [review]
Given that the `finish` code does nothing useful at this time, I do not see any point in adding it (yet).

Can you split the speaker.is_stopped code into its own PR please that this one would be dependent on?

**Comment by nielsnl68** [comment]
> Given that the `finish` code does nothing useful at this time, I do not see any point in adding it (yet).
>
The implementation of the `finish` code was very easy. and is done it the split PR#7255

> Can you split the speaker.is_stopped code into its own PR please that this one would be dependent on?

Done.

**Comment by nielsnl68** [comment]
@jesserockz this one is ready as well.

File: esphome/components/rtttl/rtttl.cpp [hunk 1]
```diff
@@ -29,6 +29,13 @@ inline double deg2rad(double degrees) {
 void Rtttl::dump_config() { ESP_LOGCONFIG(TAG, "Rtttl"); }
 
 void Rtttl::play(std::string rtttl) {
+  if (this->state_ != State::STATE_STOPPED && this->state_ != State::STATE_STOPPING) {
+    int pos = rtttl_.find(':');
```
Comment (jesserockz) [inline esphome/components/rtttl/rtttl.cpp]:
```suggestion
    int pos = this->rtttl_.find(':');
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
One minor missing prefix :see_no_evil: 
Otherwise looks good.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**